xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -109,16 +250,54 @@ private void removeServerWatcher() {
     }
   }
 
+  // if the IOException indicates we can rebuild delegate and retry start...
+  private static boolean isRetriableErrorInDelegateStart(IOException ioe) {
+    Throwable cause = ioe.getCause();
+    return cause instanceof BindException;
+  }
+
+  // if the Throwable indicates a permanent error in xDS processing
+  private static boolean isPermanentErrorFromXds(Throwable throwable) {
+    if (throwable instanceof XdsInitializationException) {
+      return true;
+    }
+    if (throwable instanceof StatusException) {
+      StatusException statusException = (StatusException) throwable;
+      Status.Code code = statusException.getStatus().getCode();
+      return EnumSet.of(
+              Status.Code.INTERNAL,
+              Status.Code.INVALID_ARGUMENT,
+              Status.Code.FAILED_PRECONDITION,
+              Status.Code.PERMISSION_DENIED,
+              Status.Code.UNAUTHENTICATED)
+          .contains(code);
+    }
+    return false;
+  }
+
+  private void startRetryTaskCleanup() {
+    if (currentServingState == ServingState.ENTER_SERVING && startRetryTask != null) {
```

Comment:
Should it be unconditional for `currentServingState`?

Reply:
probably...